### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **Bug: Inconsistent HTML tag validation regex**

The regex patterns for HTML tag validation are inconsistent between `validateTodoCreate` and `validateTodoUpdate`:
- Title uses: `/<[^>]*>/g` (matches `<script>`, ``, etc.)
- Description uses: `/<.*>/g` (matches `<script>alert("xss")</script>` but allows ``)

This creates security inconsistencies. Use the same regex pattern for both fields.

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
+  if (!id || !/^\d+$/.test(id)) {
+    return res.status(400).json({
+      error: 'Invalid ID format',
+      details: [{ field: 'id', message: 'ID must be a positive integer', value: id }]
+    });
+  }
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:109]: **Bug: ID validation allows "0" which may be invalid**

The regex `^\d+$` allows ID "0" which may not be valid depending on your ID generation strategy. Consider changing to `^[1-9]\d*$` to only allow positive integers starting from 1.

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Missing Error Handling**

There's no try-catch block around `todoStore.create()`. If the store method throws an exception, the server will crash. Consider wrapping this in a try-catch block to provide proper error handling.

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:61]: **Missing Test Coverage**

This comment highlights missing test coverage for description HTML validation. Consider adding a test like:

```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

File: src/middleware/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    it('should reject non-numeric ID', async () => {
+      const response = await request(app).get('/abc');
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'id',
+        message: 'ID must be a positive integer',
+        value: 'abc'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:119]: **Missing Test Coverage for Edge Case**

Add a test for the ID "0" edge case to ensure it behaves as expected:

```typescript
it('should handle ID "0" correctly', async () => {
  const response = await request(app).get('/0');
  // Define expected behavior - should this be valid or invalid?
  expect(response.status).toBe(/* 200 or 400 based on business rules */);
});
```

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Review Summary

This PR adds comprehensive input validation to the Todo API, which is a significant improvement for security and data integrity. The implementation is well-structured and includes good test coverage.

### ✅ **Strengths:**
- **Comprehensive validation**: Covers all required fields with appropriate constraints
- **Security focus**: Includes HTML tag prevention for XSS protection
- **Structured error responses**: Consistent error format with field-specific messages
- **Good test coverage**: 15+ test cases covering various scenarios
- **Documentation**: Updated README with validation rules and error format
- **Type safety**: Well-defined interfaces and types

### ⚠️ **Issues Found:**
1. **Security Bug**: Inconsistent HTML tag validation regex patterns between create/update
2. **Edge Case Bug**: ID validation allows "0" which may be invalid
3. **Error Handling**: Missing try-catch in POST route could crash server
4. **Test Gaps**: Missing tests for description HTML validation and ID "0" edge case

### 📋 **Recommendations:**
- Fix the regex inconsistency for HTML validation
- Clarify business rules for ID "0" and update validation accordingly
- Add error handling for store operations
- Complete test coverage for missing scenarios
- Consider adding input sanitization in addition to validation

### **Breaking Changes Note:**
The error response format change is properly documented and is a good improvement for API consumers.

Overall, this is a solid implementation that significantly improves the API's robustness. The identified issues are fixable and don't detract from the overall quality of the work.
